lib/SILOptimizer/Mandatory/DefiniteInitialization.cpp

```diff
@@ -783,6 +783,162 @@ void LifetimeChecker::diagnoseInitError(const DIMemoryUse &Use,
     diagnose(Module, TheMemory.getLoc(), diag::variable_defined_here, isLet);
 }
 
+/// Adds a hop_to_executor(self) instruction after the built-in instruction
+/// that initializes the actor with the runtime system.
+/// \param ctor the actor's constructor decl.
+/// \param entryBB the entry-block to the actor's constructor
+static void injectHopAfterBuiltinInit(ConstructorDecl *ctor,
+                                      SILBasicBlock *entryBB) {
+  SILBuilder B(entryBB);
+  SILInstruction *Inst = nullptr;
+
+  // find the call to the built-in initializeDefaultActor in the entry
+  for (auto BBI = entryBB->begin(), E = entryBB->end(); BBI != E; ++BBI) {
+    Inst = &*BBI;
+    if (auto *builtin = dyn_cast<BuiltinInst>(Inst)) {
+      if (auto kind = builtin->getBuiltinKind()) {
+        if (kind.getValue() == BuiltinValueKind::InitializeDefaultActor) {
+          // builder should point to the instruction after the built-in
+          ++BBI;
+          B.setInsertionPoint(BBI);
+          break;
+        }
+      }
+    }
+  }
+
+  if (!B.hasValidInsertionPoint())
+    llvm::report_fatal_error("did not find InitializeDefaultActor built-in!");
+
+  // get the already-borrowed `self` value from the built-in's operand.
+  assert(Inst->getNumOperands() == 1
+          && "unexpected number of operands for InitializeDefaultActor");
+  auto &firstOpr = Inst->getAllOperands().front();
+
+  // insert the hop_to_executor
+  B.createHopToExecutor(ctor, firstOpr.get(), false);
+}
+
+void LifetimeChecker::injectActorHops() {
+  auto ctor = TheMemory.isActorInitSelf();
+
+  // Must be `self` within an actor's designated initializer.
+  if (!ctor)
+    return;
+
+  // If the initializer has restricted use of `self`, then no hops are needed.
+  if (isRestrictedActorInitSelf())
+    return;
+
+  // Even if there are no stored properties to initialize, we still need a hop.
+  if (TheMemory.getNumElements() == 0)
+    return injectHopAfterBuiltinInit(ctor,F.getEntryBlock());
+
+  // A mapping from initializing blocks to their in-availability set.
+  using Map = llvm::SmallDenseMap<SILBasicBlock*, AvailabilitySet, 8>;
+
+  /////
+  // Step 1: Find initializing blocks, which are blocks that contain a store
+  // to TheMemory that fully-initializes it, and build the Map.
+
+  Map initializingBlocks;
```

Comment:
This doesn't need to be a map. It can be an array of (SILBasicBlock,  AvailabilitySet) tuples.
This also solves the problem of iterating over this data structure: iterating over a  pointer hash map is very dangerous because it can be easily introducing non-deterministic behavior.